### PR TITLE
docs(notebook): add orientation and xarray/Zarr foundations

### LINKED ISSUE #825: Establish the canonical xarray/Dask SPI/SPEI calculation path
## Context

The #822 starting script computes NumPy APIs inside explicit `dask.array.map_blocks` wrappers, while the current public typed API already accepts xarray `DataArray` inputs and uses `xr.apply_ufunc(..., dask="parallelized")`. The reader-facing workflow must choose one primary path, reconcile the notebook and helper implementation, and demonstrate real spatial parallel work without creating another execution framework.

## Goal

Establish one canonical, lazy xarray/Dask SPI/SPEI calculation path for the #822 sample.

## Scope

- Select and document the reader-facing public xarray path or retain explicit `map_blocks` only when its teaching value and contract are demonstrated.
- Reconcile `scripts/e2e_with_dask.ipynb`, `scripts/end_to_end_example.py`, and their regression coverage so they execute the same calculation path.
- Compute SPI and SPEI using the agreed Timescale, distributions, dates, Calibration Period, precipitation, and actual PET.
- Preserve labeled dimensions/coordinates and Dask laziness before a deliberate materialization or write.
- Show multiple spatial blocks while retaining one full time chunk per location.

## Out of scope

- A new generic Dask executor, workflow engine, or changes to CLI multiprocessing.
- Distributed-cluster deployment or a benchmark suite.
- New climate-index algorithms or changes to existing scientific calculations.

## Relevant repository artifacts

- `scripts/e2e_with_dask.ipynb`
- `scripts/end_to_end_example.py`
- `tests/test_e2e_with_dask.py`
- `src/climate_indices/typed_public_api.py`
- `src/climate_indices/xarray_adapter.py`
- `src/climate_indices/indices.py`
- `docs/adr/0001-dual-numpy-xarray-api.md`
- `docs/adr/0002-multiprocessing-cli-dask-xarray.md`
- `docs/adr/0003-dask-time-dimension-single-chunk.md`
- `docs/xarray_compatibility.md`
- #191, #326, #759, and #822

## Implementation considerations

The xarray API is beta, but it is the existing supported path for labeled Dask-backed SPI/SPEI and preserves metadata/coordinates. Its SPEI alignment behavior is an inner join with warning, so exact prepared inputs remain an upstream contract rather than an assumption. Explain that Dask supplies concurrent spatial tasks; worker count alone is not evidence of parallelism.

## Acceptance criteria

- [ ] The selected calculation path calls existing public climate-index APIs and has one clearly explained contract.
- [ ] SPI and SPEI receive scientifically correct inputs and agreed parameters from the #822 input contract.
- [ ] The calculated arrays retain expected `time`, `lat`, and `lon` coordinates and remain Dask-backed until a deliberate execution boundary.
- [ ] The time dimension is a single chunk; the example has multiple spatial blocks visible in the chunk/task layout.
- [ ] Notebook and companion implementation no longer diverge, and focused regression coverage exercises the path users run.

## Dependencies

Blocked by #823 and #824. Native GitHub dependencies are authoritative.

## Verification

Run focused workflow checks; inspect chunks/task graph; compare representative grid cells with direct stable NumPy API calculations before persisting output.

### LINKED ISSUE #828: Build notebook orientation and xarray/Zarr foundations
## Context

#822 is a teaching deliverable for readers new to xarray, Dask, Zarr, SPI/SPEI, and the package APIs. The current starting notebook has code cells only, stale parameters, placeholder-style paths, and no reader orientation. Existing small notebooks demonstrate individual xarray API calls but not this end-to-end workflow.

## Goal

Build the #822 notebook's setup, orientation, xarray/Zarr data-model, and scientific input-validation sections.

## Scope

- Establish portable environment/setup commands, working-directory expectations, one-time preparation instructions, and a clear notebook location decision.
- Add an opening workflow diagram, scientific question, outcomes, and immediate definitions of SPI, SPEI, PET, Timescale, Periodicity, and Calibration Period.
- Teach `Dataset` versus `DataArray`, labeled dimensions/coordinates/attributes, lazy `xr.open_zarr`, and distinct roles for xarray, Zarr, and Dask.
- Walk through concise, executable inspection and validation of variable names, dates, units, dimensions, coordinates, masks, chunks, and the agreed #822 input contract.
- Explain that prepared Zarr is the normal tutorial entry point and source preparation is an optional reproducible prerequisite.

## Out of scope

- Dask execution mechanics, calculation implementation, persisted-output plotting, and final troubleshooting sections.
- A generic data-requirements manual; connect evidence to the project discovery issue instead.
- A second source dataset or invented station observations.

## Relevant repository artifacts

- `scripts/e2e_with_dask.ipynb`
- `scripts/prepare_e2e_inputs.py`
- `scripts/end_to_end_example.py`
- `notebooks/xarray_getting_started.ipynb`
- `docs/xarray_migration.rst`
- `docs/xarray_compatibility.md`
- `docs/index.rst`
- `data/e2e/manifest.json`
- #401 and #822

## Implementation considerations

Use project terminology: Timescale in prose (`scale` in code), Periodicity, and Calibration Period. State that xarray labeling, Zarr storage, and Dask execution are complementary rather than interchangeable. Do not present attribute relabeling as physical conversion or make a sample result an independently validated drought assessment.

## Acceptance criteria

- [ ] A reader can set up the documented environment and open the prepared store without machine-specific paths or hidden state.
- [ ] Markdown introduces the workflow and scientific terms before dependent code cells.
- [ ] Executable cells inspect the Zarr dataset, variables, dimensions, coordinates, dates, units, masks, and chunks, then enforce the agreed input checks.
- [ ] The notebook clearly separates required correctness checks from optional performance tuning.
- [ ] Source attribution, input identity, regeneration instructions, and generated-data policy are discoverable.

## Dependencies

Blocked by #823 and #824. Native GitHub dependencies are authoritative.

## Verification

Execute the foundations from a fresh kernel against prepared local data and confirm every displayed fact matches the manifest/store metadata.

### PR DESCRIPTION
**Pull request by monocongo**
Resolves #825
Resolves #828

## Summary

- The e2e teaching notebook is now the authoritative artifact: it moved to `notebooks/zarr_dask_spi_spei.ipynb`, imports only `climate_indices`, xarray, dask, and matplotlib, and performs validation, calculation (public `spi()`/`spei()`), atomic Zarr write, reopen, and plots in visible cells. `scripts/end_to_end_example.py` is retired; its preparation helpers moved into `scripts/prepare_e2e_inputs.py`, their only remaining consumer.
- Foundations for the #822 notebook: orientation and workflow diagram, terminology (SPI, SPEI, PET, Timescale, Periodicity, Calibration Period), environment/setup and one-time input preparation, source attribution for the reduced nClimGrid sample, configuration and manifest display, the xarray/Zarr/Dask data model, a lazy open with inspection, and the input contract as executable checks (complete monthly time axis, calibration years inside the data range, exact precip/PET alignment, mm units, matching masks, zero-versus-missing, single time chunk, manifest agreement with the store).
- Tests: the notebook-path test executes the notebook's code cells against a tiny synthetic prepared store (data dir via `CLIMATE_INDICES_E2E_DATA`, with `current/` layout and a manifest) instead of importing a script. Prep-validation tests moved to `tests/test_prepare_e2e_inputs.py`. Missing prepared inputs or manifest raise actionable errors naming the regeneration command.

## Verification

- Environment: macOS arm64, Python 3.13.13, `uv sync --group dev`.
- Inputs: `uv run --with h5py --with zarr scripts/prepare_e2e_inputs.py` → `data/e2e/current/manifest.json` records 444 × 38 × 87, 1980-01-01 through 2016-12-01, chunks `[444, 10, 10]`.
- Fresh-kernel execution: `uv run jupyter nbconvert --execute --to notebook --output-dir <scratchpad> notebooks/zarr_dask_spi_spei.ipynb` completed in 10.3 s; every displayed fact matches the manifest and store metadata: dims, dates, units, 360 calibration observations per cell, 1474 of 3306 all-missing cells, 0 partially missing cells, 7051 zero-precipitation observations, and 36 = 4×9 spatial blocks.
- Negative check: pointing `CLIMATE_INDICES_E2E_DATA` at an empty directory raises `FileNotFoundError: Prepared inputs not found: ... Generate them once with: uv run --with h5py --with zarr scripts/prepare_e2e_inputs.py`.
- Tests: `uv run pytest` → 1320 passed, 1 skipped, 395 deselected. `uv run pytest tests/test_e2e_with_dask.py tests/test_prepare_e2e_inputs.py -q` passes.
- Lint and types: `uv run ruff check src/ tests/`, `uv run ruff format --check src/ tests/`, and `uv run mypy src/` are clean. Pre-existing, unrelated: `ruff format --check scripts/` would reformat `scripts/compare_versions.py`, which this branch does not touch.
- The notebook is committed with outputs cleared; `grep -n '/Users/'` finds nothing; `git status` shows no `data/` files.

## Notes

- Three commits: decouple from the retired script, move under `notebooks/`, then foundations.
- The single-`time`-chunk check asserts one chunk rather than the literal `(444,)` so the synthetic-store test remains meaningful; the executed notebook displays `(444,)` on the real inputs.

## Out of scope

- Dask client, graph, and compute/persist mechanics (#829).
- Output metadata and round trip (#826).
- Maps and time series (#830, #831).
- Docs index link, troubleshooting, and the CI smoke command (#833).

## Summary by Sourcery

Establish the xarray/Zarr/Dask teaching notebook as the canonical end-to-end SPI/SPEI workflow and strengthen its preparation, validation, publication, and regression coverage.

New Features:
- Add an authoritative end-to-end teaching notebook covering xarray, Zarr, Dask, SPI/SPEI calculation, validation, persistence, reopening, and visualization.
- Document the terminology, data model, setup, provenance, configuration, and executable input contract for the prepared climate dataset.

Bug Fixes:
- Make prepared-input validation reject invalid time axes, units, coordinate alignment, missingness, chunking, and manifest mismatches with actionable errors.
- Preserve previously published output when an atomic result publication is interrupted or fails.

Enhancements:
- Move input preparation helpers into the standalone preparation script and retire the former end-to-end script and notebook.
- Update xarray compatibility guidance to reference the new teaching notebook and its test coverage.

Documentation:
- Add user-facing guidance and source attribution for the reduced nClimGrid example and its Zarr/xarray/Dask workflow.

Tests:
- Execute the notebook against synthetic prepared stores and expand coverage for validation failures, actionable missing-input errors, plotting, and interrupted publication.

## Summary by CodeRabbit

- **Documentation**
  - Added a notebook-based guide for lazy xarray/Dask SPI and SPEI workflows.
  - Updated compatibility guidance with chunking, alignment, and parallel execution recommendations.
  - Replaced references to the previous end-to-end examples.

- **New Features**
  - Added comprehensive validation for time ranges, units, coordinates, calibration periods, and prepared data stores.
  - Added atomic result writing and visual output examples to the workflow notebook.

- **Tests**
  - Expanded coverage for validation, complete SPI/SPEI execution, failure handling, and output preservation.

### CODE REVIEW

File: notebooks/zarr_dask_spi_spei.ipynb [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9544de71-134caaba-77f3-4c69-b8bf-b0987e6ef3b2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "precip, pet = ds[\"precip\"], ds[\"pet\"]\n",
+    "precip_mask, pet_mask = precip.isnull().any(\"time\"), pet.isnull().any(\"time\")\n",
+    "if not precip_mask.equals(pet_mask):\n",
```
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/zarr_dask_spi_spei.ipynb]: Fixed in ccf19eae: the check now compares `precip.isnull().equals(pet.isnull())` across the full time axis before any reduction, so per-timestamp mismatches are rejected. A regression test (`test_notebook_rejects_mismatched_missingness`) covers the January-precip/February-PET case.

File: notebooks/zarr_dask_spi_spei.ipynb [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+    "        \"Precipitation and PET masks must match; SPEI would otherwise combine present values with missing ones.\",\n",
+    "        argument_name=\"pet\",\n",
+    "        argument_value=\"mask differs from precip\",\n",
+    "        valid_values=\"identical masks\",\n",
+    "    )\n",
+    "\n",
+    "all_missing = precip.isnull().all(\"time\")\n",
+    "partially_missing = precip.isnull().any(\"time\") & ~all_missing\n",
+    "print(f\"All-missing cells: {int(all_missing.sum())} of {all_missing.size}\")\n",
+    "print(f\"Partially missing cells: {int(partially_missing.sum())}\")\n",
```
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/zarr_dask_spi_spei.ipynb:467]: Fixed in ccf19eae: partially missing cells now raise `InvalidArgumentError` instead of only being counted, with a regression test (`test_notebook_rejects_partially_missing_cells`). The prepared 1980-2016 sample still reports 0 partially missing cells, so behavior on real inputs is unchanged.

File: llms-full.txt [hunk 3]
```diff
@@ -432,12 +432,11 @@ no coordinates.
...
 - Keep Dask chunks spatial when possible and leave `time` as one chunk before
   calling index functions.
-- The canonical lazy xarray/Dask SPI/SPEI workflow is
-  `scripts/end_to_end_example.py` with `scripts/e2e_with_dask.ipynb`: the public
-  typed API on Dask-backed DataArrays (`xr.apply_ufunc(..., dask="parallelized")`),
-  one full time chunk with spatial chunks driving task parallelism, and
-  precipitation/PET exact-aligned at preparation time so SPEI never relies on
-  coordinate intersection.
+- The canonical lazy xarray/Dask SPI/SPEI workflow is the teaching notebook
+  `notebooks/zarr_dask_spi_spei.ipynb`: the public typed API on Dask-backed DataArrays
```
[the remark this answers is not in the document]
Comment (monocongo) [reply llms-full.txt:436]: Out of scope here: #833 owns the docs index link, troubleshooting, and the CI smoke command. Running this notebook in the documented CI command requires the prepared store in CI, which is exactly what #833 designs; adding the notebook to the existing command now would make that documented validation fail without the optional one-time preparation.

File: notebooks/zarr_dask_spi_spei.ipynb [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+    "float32_encoding = {\"dtype\": \"float32\"}\n",
+    "ds_output.to_zarr(\n",
+    "    tmp_path,\n",
+    "    mode=\"w\",\n",
+    "    zarr_format=2,\n",
+    "    consolidated=True,\n",
+    "    encoding={spi_name: float32_encoding, spei_name: float32_encoding},\n",
+    ")\n",
+    "shutil.rmtree(final_output_zarr, ignore_errors=True)\n",
+    "tmp_path.rename(final_output_zarr)"
```
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/zarr_dask_spi_spei.ipynb]: Out of scope here: #826 owns output metadata and the round trip (including publication semantics). The write-beside-and-swap logic came from the #825 canonical path (PR #844) and is unchanged by this foundations work; I have left it as-is and flagged the backup/restore suggestion for #826.

File: notebooks/zarr_dask_spi_spei.ipynb [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9544de71-134caaba-77f3-4c69-b8bf-b0987e6ef3b2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "precip, pet = ds[\"precip\"], ds[\"pet\"]\n",
+    "precip_mask, pet_mask = precip.isnull().any(\"time\"), pet.isnull().any(\"time\")\n",
```
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/zarr_dask_spi_spei.ipynb]: Addressed in ccf19eae: missingness is now compared at every timestamp (`precip.isnull().equals(pet.isnull())`) and partially missing cells raise `InvalidArgumentError`, so neither pattern reaches the fitting calls. Regression tests cover both.

File: notebooks/zarr_dask_spi_spei.ipynb [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+    "# footprint at the float32 precision the float32 mm inputs actually carry.\n",
+    "float32_encoding = {\"dtype\": \"float32\"}\n",
+    "ds_output.to_zarr(\n",
+    "    tmp_path,\n",
+    "    mode=\"w\",\n",
+    "    zarr_format=2,\n",
+    "    consolidated=True,\n",
+    "    encoding={spi_name: float32_encoding, spei_name: float32_encoding},\n",
+    ")\n",
+    "shutil.rmtree(final_output_zarr, ignore_errors=True)\n",
```
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/zarr_dask_spi_spei.ipynb]: Out of scope here: #826 owns output metadata and the round trip, including publication semantics. This write-and-swap code is unchanged from the #825 canonical path (PR #844); I have flagged the generation-retention point for #826.

File: notebooks/zarr_dask_spi_spei.ipynb [hunk 7]
```diff
@@ -0,0 +1,701 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b7c146d1-2ecc08c9-b40e-4c40-91e6-ccee11c855ed",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Write beside the target and swap on success so a failed run leaves a\n",
+    "# previously completed store intact.\n",
+    "tmp_path = final_output_zarr.with_name(final_output_zarr.name + \".tmp\")\n",
```
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/zarr_dask_spi_spei.ipynb]: Out of scope here: concurrent-invocation safety belongs with the output/publication semantics in #826. This notebook is a single-user tutorial workflow with a local output path inherited from PR #844.

File: docs/xarray_compatibility.md [hunk 8]
```diff
@@ -48,11 +48,10 @@ no coordinates.
...
 - Keep Dask chunks spatial when possible and leave `time` as one chunk before
   calling index functions.
-- The canonical lazy xarray/Dask SPI/SPEI workflow is
-  `scripts/end_to_end_example.py` with `scripts/e2e_with_dask.ipynb`: the public
-  typed API on Dask-backed DataArrays (`xr.apply_ufunc(..., dask="parallelized")`),
-  one full time chunk with spatial chunks driving task parallelism, and
-  precipitation/PET exact-aligned at preparation time so SPEI never relies on
-  coordinate intersection.
+- The canonical lazy xarray/Dask SPI/SPEI workflow is the teaching notebook
+  `notebooks/zarr_dask_spi_spei.ipynb`: the public typed API on Dask-backed DataArrays
```
[the remark this answers is not in the document]
Comment (monocongo) [reply docs/xarray_compatibility.md:52]: Out of scope here: #833 owns the docs index, troubleshooting, and the CI smoke command, including making prepared-store-dependent notebook execution part of CI. Until that lands, the notebook-path test executes the notebook's own cells against a synthetic store so the contract is covered without the optional download.

File: notebooks/zarr_dask_spi_spei.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/zarr_dask_spi_spei.ipynb]: Fixed in a98d8bc1: the write-and-swap cell now stages each run under a private, uuid-suffixed name and renames the previous generation aside (fast, atomic) instead of deleting it before the rename. An interruption between the two renames now leaves the previous generation recoverable at its backup path instead of the documented output being simply absent. This supersedes my earlier reply deferring the fix to #826 — the write-beside-and-swap flow is no longer unchanged from PR #844. Regression test: test_interrupted_publish_preserves_previous_generation (tests/test_e2e_with_dask.py).

File: notebooks/zarr_dask_spi_spei.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/zarr_dask_spi_spei.ipynb]: Fixed in a98d8bc1: each run now stages under a unique `.climate_indices_output.zarr.staging-<uuid>` name instead of the shared `.tmp` path, so a concurrent invocation can no longer delete or publish over another run's in-progress output. This supersedes my earlier reply deferring the fix to #826.

File: notebooks/zarr_dask_spi_spei.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/zarr_dask_spi_spei.ipynb]: Correction to the learning recorded on this thread: the write-beside-and-swap publication flow is no longer unchanged from PR #844. Fixed in a98d8bc1 — it now stages under a unique name per run and swaps by renaming the previous generation aside instead of deleting it first, addressing this finding and the two duplicate tessl-reviewer findings on the same cell. Please disregard the prior 'unchanged by the foundations work' learning for this file.

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/861
- **Lanes run:** 4 independent lanes (python-reviewer: correctness of the moved/inlined validation logic; pr-test-analyzer: test coverage quality after the script→notebook split; architect: maintainability of inlining the pipeline and the `_validate_monthly_time` duplication; general-purpose: scientific terminology, ADR references, and data-attribution accuracy), plus triage of the standing `tessl-reviewer` bot review (4 findings) and a separate `coderabbitai` review, both already on the PR.
- **Validation:** `uv run pytest -q` → 1324 passed, 0 failed, 1 skipped (pre-existing, unrelated `h5py` optional-dependency skip). `uv run ruff check src/ tests/ notebooks/zarr_dask_spi_spei.ipynb` → clean. `uv run ruff format --check src/ tests/` → clean (69 files; `scripts/compare_versions.py` is pre-existing/unrelated drift already disclosed in the PR description). `uv run mypy src/` → no issues.
- **Fix commits:**
  - [`a98d8bc1` fix(notebook): make output publish crash-safe and collision-free](https://github.com/monocongo/climate_indices/commit/a98d8bc1)
  - [`0fc9be61` fix(notebook): close the Dask client after the Zarr write](https://github.com/monocongo/climate_indices/commit/0fc9be61)
  - [`d53ba2ce` docs(notebook): soften the nClimGrid provenance-doc citation](https://github.com/monocongo/climate_indices/commit/d53ba2ce)
  - [`1005e742` docs(notebook): describe the CI notebook gate's actual coverage](https://github.com/monocongo/climate_indices/commit/1005e742)

### Findings and fix status

#### [P1] Output publish has a window with no completed output on interruption — Fixed
- **Location:** `notebooks/zarr_dask_spi_spei.ipynb` (write/publish cell)
- **Origin:** `tessl-reviewer` (also raised independently by `coderabbitai`)
- **Problem:** The publish cell deleted the existing completed `climate_indices_output.zarr` before renaming the new store into place. A crash or rename failure in that interval left the documented output path absent even though a prior completed result existed; the existing preservation test only covered failures *before* that deletion.
- **Applied fix:** Stage each run under a private, uuid-suffixed name and swap by renaming the previous generation aside (fast, atomic) instead of deleting it first, so an interruption mid-publish leaves the previous generation recoverable rather than destroyed. New regression test `test_interrupted_publish_preserves_previous_generation` simulates a crash during the swap and asserts the previous generation survives. Commit `a98d8bc1`.
- **Note:** This was previously triaged in this PR's review threads as "out of scope, belongs to #826" (the code originated unchanged from PR #844). Re-opened and fixed here per explicit direction this session; the corresponding threads (including the duplicate `coderabbitai` finding, which had recorded a now-stale "unchanged" learning) were replied to and resolved with the new commit.

#### [P2] Concurrent notebook executions share and can delete each other's staging store — Fixed
- **Location:** `notebooks/zarr_dask_spi_spei.ipynb` (write/publish cell)
- **Origin:** `tessl-reviewer`
- **Problem:** Every invocation staged under the same fixed `climate_indices_output.zarr.tmp` name and unconditionally removed it before writing, so two concurrent runs against the same data root could delete or publish over each other's in-progress output.
- **Applied fix:** Each run now stages under a unique `.climate_indices_output.zarr.staging-<uuid>` name (part of the same `a98d8bc1` change as the P1 fix above).
- **Note:** Also previously deferred to #826 on this PR's review thread; fixed here alongside the P1 change, and the thread was resolved with the new commit.

#### [P2] Mask check compared missingness per-cell instead of per-timestamp; partial missingness wasn't rejected — Already fixed, thread resolved
- **Location:** `notebooks/zarr_dask_spi_spei.ipynb` (input-contract mask cell)
- **Origin:** `tessl-reviewer`
- **Problem:** An earlier revision reduced over `time` before comparing precip/PET masks (so mismatched missing months in different calendar months could pass) and only reported, rather than rejected, partially missing cells.
- **Applied fix:** Already fixed on this branch before this review, in `ccf19eae` (`precip.isnull().equals(pet.isnull())` compared at every timestamp; `partially_missing.any()` now raises). This session verified the fix against the pre-fix commit and resolved the review thread, which had a correct reply but was never marked resolved.

#### [P3] Dask distributed `Client` created in the notebook was never closed — Fixed
- **Location:** `notebooks/zarr_dask_spi_spei.ipynb` (Dask execution cell)
- **Origin:** python-reviewer lane
- **Problem:** `Client(n_workers=4, ...)` had no matching `close()`, unlike the retired script's `with Client(...):` block. A user re-running that cell (or restarting just it) accumulates worker processes across runs until the kernel restarts. Tests exclude this cell entirely, so it had no coverage of its own lifecycle.
- **Applied fix:** Added a `client.close()` cell immediately after the Zarr write (once the lazily scheduled work is materialized). Commit `0fc9be61`.

#### [P3] Plot cells had zero execution coverage even where matplotlib is installed — Fixed
- **Location:** `tests/test_e2e_with_dask.py`
- **Origin:** pr-test-analyzer lane
- **Problem:** `.plot(` cells were unconditionally excluded from every test, including the main CI job where matplotlib and `distributed` are installed dev dependencies — so a renamed variable between the calculation cells and the plot cells (a real, not hypothetical, failure mode) would go undetected until someone ran the notebook by hand.
- **Applied fix:** Added `test_notebook_plot_cells_execute`, gated on `pytest.importorskip("matplotlib")` with the `Agg` backend, that executes the full pipeline through both plot cells. The Dask `Client` cell itself remains excluded from all tests (unchanged, tracked under #829) since spinning a real multi-process cluster inside the shared test fixtures would be slow and flaky; this is a deliberate, narrower scope than fixing every gap the lane raised. Commit `0fc9be61`.

#### [P3] "Full provenance and redistribution constraints" overclaims a scoped, unrelated doc — Fixed
- **Location:** `notebooks/zarr_dask_spi_spei.ipynb` (source-attribution cell)
- **Origin:** general-purpose (scientific-accuracy) lane
- **Problem:** `docs/research/nclimgrid-acquisition-and-redistribution.md` is scoped to a different, not-yet-built feature's sample (issue #726) and never mentions this notebook's actual source (`monocongo/example_climate_indices`, pinned commit, SHA-256 checksums). Calling it "Full provenance and redistribution constraints" overstates what the linked file delivers.
- **Applied fix:** Reworded to "Background on the upstream provenance and redistribution constraints this sample inherits," matching the wording `scripts/prepare_e2e_inputs.py` already uses for the same doc. Commit `d53ba2ce`.

#### [P3] Docs call the new notebook canonical, then list a CI gate that doesn't cover it — Fixed (doc clarified); underlying gap still deferred
- **Location:** `docs/xarray_compatibility.md`
- **Origin:** `tessl-reviewer`; independently corroborated by the pr-test-analyzer lane (the exclusion rationale in the test file's own comment was also imprecise — it blamed "the minimum-dependency job," but the *main* CI job has matplotlib/distributed installed and still excludes these cells)
- **Problem:** The doc names `notebooks/zarr_dask_spi_spei.ipynb` as canonical, then gives a "notebook CI command" and points at a CI `notebooks` job that both only ever execute the three older notebooks.
- **Applied fix:** Added a note that the command and CI job cover only those three notebooks, and that the canonical notebook is instead covered by `tests/test_e2e_with_dask.py` against a synthetic store (since it needs prepared inputs that job doesn't provision). Commit `1005e742`. This thread was already resolved on the PR (deferred to #833); the doc-level clarification is additive, not a scope reopening. Also corrected the test file's own comment to stop misattributing the plot/Client exclusion to the minimum-dependency job specifically (commit `0fc9be61`).
- **Remaining:** Actually running the canonical notebook end-to-end in CI (via `scripts/prepare_e2e_inputs.py`) stays out of scope for this PR, tracked in #833, per the PR's own stated scope.

#### [P4] `_validate_monthly_time` duplicated between `scripts/prepare_e2e_inputs.py` and the notebook — Blocked (scope decision)
- **Location:** `notebooks/zarr_dask_spi_spei.ipynb`, `scripts/prepare_e2e_inputs.py:71-98`
- **Origin:** architect lane
- **Problem:** Two independent, currently-identical copies of the monthly-time-contract validation, with asymmetric test coverage (the notebook copy is exercised by only one of the four branches the script copy is tested for) — a real, if narrow, drift risk.
- **Blocked:** This is a direct consequence of the PR's stated design goal (the notebook imports only `climate_indices`, `xarray`, `dask`, `matplotlib` — no `scripts/` import). Removing the duplication cleanly would mean moving this helper into the shipped `climate_indices` package's internal module so both call sites import one implementation — an architecture/scope decision beyond a "notebook foundations" PR, and out of proportion to fix unilaterally here. Recommend a follow-up issue if the maintainer wants this consolidated.

#### [P4] Inlining `compute_indices_parallel`'s logic as loose notebook cells instead of a reusable function — Not fixed (deliberate trade-off)
- **Location:** `notebooks/zarr_dask_spi_spei.ipynb`
- **Origin:** architect lane
- **Problem:** The old script's `compute_indices_parallel(prepared, output, config)` was a real, reusable, directly-unit-testable function. It's now ~15 sequential notebook cells reachable only via `exec()`-based cell extraction in tests, which is more fragile (relies on substring matches like `"open_zarr" in source` to find the setup/calculation split) and removes the only non-notebook entry point to this pipeline.
- **Not fixed:** Deliberate and reasonable given the PR's explicit "the notebook is now the authoritative teaching artifact" goal — inlining lets a learner inspect `ds`, `time`, `manifest`, etc. between steps, which the old opaque function call didn't allow. No other caller of the assembled pipeline exists or is planned. This is a real, nameable cost (documented here) rather than a defect to fix.

### Remaining blockers
None. No P0/P1 findings remain open.
